pkgs/development/interpreters/python/cpython/3.7/default.nix

```diff
@@ -0,0 +1,183 @@
+{ stdenv, fetchurl, fetchpatch
+, glibc
+, bzip2
+, expat
+, libffi
+, gdbm
+, lzma
+, ncurses
+, openssl
+, readline
+, sqlite
+, tcl ? null, tk ? null, tix ? null, libX11 ? null, xproto ? null, x11Support ? false
+, zlib
+, callPackage
+, self
+, CF, configd
+, python-setup-hook
+# For the Python package set
+, pkgs, packageOverrides ? (self: super: {})
+}:
+
+assert x11Support -> tcl != null
+                  && tk != null
+                  && xproto != null
+                  && libX11 != null;
+with stdenv.lib;
+
+let
+  majorVersion = "3.7";
+  minorVersion = "0";
+  minorVersionSuffix = "a4";
+  pythonVersion = majorVersion;
+  version = "${majorVersion}.${minorVersion}${minorVersionSuffix}";
+  libPrefix = "python${majorVersion}";
+  sitePackages = "lib/${libPrefix}/site-packages";
+
+  buildInputs = filter (p: p != null) [
+    zlib bzip2 expat lzma libffi gdbm sqlite readline ncurses openssl ]
+    ++ optionals x11Support [ tcl tk libX11 xproto ]
+    ++ optionals stdenv.isDarwin [ CF configd ];
+
+in stdenv.mkDerivation {
+  name = "python3-${version}";
+  pythonVersion = majorVersion;
+  inherit majorVersion version;
+
+  inherit buildInputs;
+
+  src = fetchurl {
+    url = "https://www.python.org/ftp/python/${majorVersion}.${minorVersion}/Python-${version}.tar.xz";
+    sha256 = "0na11gyqq09zklh4swbzd24irhspcc60iwk11nq4qa2klfw9963y";
+  };
+
+  NIX_LDFLAGS = optionalString stdenv.isLinux "-lgcc_s";
+
+  # Determinism: The interpreter is patched to write null timestamps when compiling python files.
+  # This way python doesn't try to update them when we freeze timestamps in nix store.
+  DETERMINISTIC_BUILD=1;
+  # Determinism: We fix the hashes of str, bytes and datetime objects.
+  PYTHONHASHSEED=0;
+
+  prePatch = optionalString stdenv.isDarwin ''
+    substituteInPlace configure --replace '`/usr/bin/arch`' '"i386"'
+    substituteInPlace configure --replace '-Wl,-stack_size,1000000' ' '
+  '';
+
+  patches = [
+    ./no-ldconfig.patch
+  ];
+
+  postPatch = ''
+    # Determinism
+    substituteInPlace "Lib/py_compile.py" --replace "source_stats['mtime']" "(1 if 'DETERMINISTIC_BUILD' in os.environ else source_stats['mtime'])"
```

Comment:
I did not go through [pep-0552](https://www.python.org/dev/peps/pep-0552/) intensively, but could not use the fact that it offers the possibility to have timestamp-less .pyc files and replace this hack?

something like:

```bash
substituteInPlace "Lib/py_compile.py" --replace "invalidation_mode=PycInvalidationMode.TIMESTAMP):" "invalidation_mode=PycInvalidationMode.CHECKED_HASH):"
```

and remove `DETERMINISTIC_BUILD` env var ? (this wold require some testing…)

Reply:
yep, that's on the todo list. We will need to modify some of this patching.

Reply:
```bash
substituteInPlace "Lib/py_compile.py" --replace "invalidation_mode=PycInvalidationMode.TIMESTAMP):" "invalidation_mode=(PycInvalidationMode.CHECKED_HASH if 'DETERMINISTIC_BUILD' in os.environ else PycInvalidationMode.TIMESTAMP)):"
```
 would probably be better